sdk/core/azure-core/azure/core/rest/_rest.py

```diff
@@ -0,0 +1,376 @@
+                continue
+            self.headers.setdefault(name, value)
+
+    @property
+    def content(self):
+        # type: (...) -> Any
+        """Get's the request's content
+
+        :return: The request's content
+        :rtype: any
+        """
+        return self._data or self._files
+
+    def __repr__(self):
+        # type: (...) -> str
+        return "<HttpRequest [{}], url: '{}'>".format(
+            self.method, self.url
+        )
+
+    def __deepcopy__(self, memo=None):
+        try:
+            request = HttpRequest(
+                method=self.method,
+                url=self.url,
+                headers=self.headers,
+            )
+            request._data = copy.deepcopy(self._data, memo)
+            request._files = copy.deepcopy(self._files, memo)
+            return request
+        except (ValueError, TypeError):
+            return copy.copy(self)
+
+    def _to_pipeline_transport_request(self):
+        return to_pipeline_transport_request_helper(self)
+
+    @classmethod
+    def _from_pipeline_transport_request(cls, pipeline_transport_request):
+        return from_pipeline_transport_request_helper(cls, pipeline_transport_request)
+
+class _HttpResponseBase(object):  # pylint: disable=too-many-instance-attributes
+
+    def __init__(self, **kwargs):
+        # type: (Any) -> None
+        self.request = kwargs.pop("request")
+        self._internal_response = kwargs.pop("internal_response")
+        self.status_code = None
+        self.headers = {}  # type: HeadersType
+        self.reason = None
+        self.is_closed = False
+        self.is_stream_consumed = False
+        self._num_bytes_downloaded = 0
+        self.content_type = None
+        self._json = None  # this is filled in ContentDecodePolicy, when we deserialize
+        self._connection_data_block_size = None  # type: Optional[int]
+        self._content = None  # type: Optional[bytes]
+
+    @property
+    def url(self):
+        # type: (...) -> str
+        """Returns the URL that resulted in this response"""
+        return self.request.url
+
+    @property
+    def encoding(self):
+        # type: (...) -> Optional[str]
+        """Returns the response encoding. By default, is specified
+        by the response Content-Type header.
+        """
+        try:
+            return self._encoding
+        except AttributeError:
+            return get_charset_encoding(self)
+
+    @encoding.setter
+    def encoding(self, value):
+        # type: (str) -> None
+        """Sets the response encoding"""
+        self._encoding = value
+
+    @property
+    def text(self):
+        # type: (...) -> str
+        """Returns the response body as a string"""
+        encoding = self.encoding
+        if encoding == "utf-8" or encoding is None:
+            encoding = "utf-8-sig"
+        return self.content.decode(encoding)
+
+    @property
+    def num_bytes_downloaded(self):
```

Comment:
Please hide it. There is no reliable way to return the bytes.

Reply:
All I do in num_bytes_downloaded is I get. the length of the chunk before I send it to users, I feel like that's p reliable, why are you thinking it's not reliable? sorry if I'm missing something

Reply:
in compression scenario. Which num is the correct? Before decompression or after?

Reply:
you're right that seems fair, i can remove for now